### ISSUE TITLE
video.currentTime is not precise enough to indentify a frame

### ISSUE THREAD

**Issue by mzur** [biigle/core#433]
Apparently, `video.currentTime` can be modified/rounded by the browser [for security reasons](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/currentTime#usage_notes). This means that the value returned by `currentTime` is not precise enough to identify the video frame. The following observation confirms this issue: If the video is played and paused and then `video.currentTime = video.currentTime` is set, **the displayed video frame is updated** most of the time! Try it in [this fiddle](https://jsfiddle.net/ek6wa547/).

The issue that frame-accurate handling of videos is not really possible in the browser is discussed [for a long time](https://www.w3.org/2019/Talks/TPAC/frame-accurate-sync). I don't know how we can handle this, yet, but it might become complicated.

**Comment by mzur** [comment]
See [this discussion](https://github.com/cvisionai/tator/discussions/658) for how a switch to frame numbers instead of times could be accomplished.

**Comment by mzur** [comment]
Also related: https://github.com/w3c/media-and-entertainment/issues/4

**Comment by mzur** [comment]
PR #444 implements a workaround where the video frame is updated whenever the video is paused. The updated frame exactly shows the frame corresponding to `currentTime` so annotations placed at this time will always resolve back to the correct frame. The downside is that the video jumps one frame forward or backward whenever the video is paused. Considering the very complex alternatives, I can live with that. I'll leave this open since this issue should be fixed eventually (maybe with updated web APIs?) but I'll lower the priority, as it is no longer pressing.

**Comment by mzur** [comment]
Here is an example how a video could be rendered frame by frame using the Web Codecs API: https://stackoverflow.com/a/32708998/1796523

**Comment by mzur** [comment]
Here is a project that renders videos using the WebCodecs API: https://github.com/aegir-assembly/omni-clip
It uses FFMpeg (WASM) to demux the video files. This could be a solution for our support of many different containers/codecs. WHile the whole of FFMpeg is very large, maybe we can only compile a subset of it that is required for demuxing.

Another thing to consider is that with WebCodecs we can only support remote files with CORS enabled (similar to the image annotation tool).

**Comment by davidsean** [comment]
In order to address some of the issues, I wrote a [community script](https://github.com/davidsean/community-resources/tree/feature/snap_annotations)  that will "snap" time annotations to an exact frame (at least with floating point accuracy).

## video @ 5 fps
This is a simple case, and easy to get a timestamps between frames.
I manually tried to make an annotation at the 3.0s mark (jumping individual frames). I saw that the query param was wrong, but didn't care. Using the API, I confirmed that the annotation time ([confusingly called](https://biigle.de/doc/api/index.html#api-VideoAnnotations-UpdateVideoAnnotation) `frame`) was actually between frames. My script was able to snap the annotation times to the exact time of the previous frame.

At this point I consider the annotation correct, and I can visually test the annotation front-end as well as the largo thumbnail. When the annotation front-end is reloaded and I click on the annotation to jump to the frame, I see that the query param is `&t=30000` and that the frame visually is at t=3.0s and frame 15--- so far so good! However, the largo thumbnail is NOT regenerated, so I cannot comment on it's accuracy (I will wait a bit in case a worker was not available, I don't know if the API triggers rebuilding the thumbnail.

ISSUE 1: rebuild the largo thumbnail.
ISSUE 2: (possibly?) fix the the largo thumbnail so the correct frame is displayed.

## video @ 29.97fps (or exactly 30000./1001.)
using a frame_rate of `30000./1001` is the annoying case. I added the same annotations at t=3s the best I could (same as before) and "snapped" that annotation to the exact timestamp using the script. This time, the time is snapped to `'frames': [2.9696333333333333]` (according to the API). That's nice, the API allows me to use more precision that the annotation front-end!

However, when I try too see the annotation the wrong frame is shown.
![image](https://github.com/user-attachments/assets/b61a5b91-ff90-49d4-b621-3b01ffd942bf)

The query param is `&t=29696` (the truncated time caused the frame to round down) which loaded frame 88 (instead of 89).

ISSUE 4: annotation front-end should use the timestamp from the annotation object when available (instead of the query-parameter).

TLDR:
I generated videos that display their own frame number and time stamp. By using the API I can define an annotation with high precision. This allowed to independently test largo and the annotation front-end to see how they render a frame. There are a few different independent on both sides issues that could be addressed.
